zebra-consensus/src/chain.rs

```diff
@@ -0,0 +1,178 @@
+//! Chain state updates for Zebra.
+//!
+//! Chain state updates occur in multiple stages:
+//!   - verify blocks (using `BlockVerifier` or `CheckpointVerifier`)
+//!   - update the list of verified blocks on disk
+//!   - create the chain state needed to verify child blocks
+//!   - choose the best tip from all the available chain tips
+//!   - update the mature chain state on disk
+//!   - prune orphaned side-chains
+//!
+//! Chain state updates are provided via a `tower::Service`, to support
+//! backpressure and batch verification.
+
+#[cfg(test)]
+mod tests;
+
+use crate::checkpoint::CheckpointVerifier;
+
+use futures_util::FutureExt;
+use std::{
+    error,
+    future::Future,
+    pin::Pin,
+    sync::Arc,
+    task::{Context, Poll},
+};
+use tower::{buffer::Buffer, Service, ServiceExt};
+
+use zebra_chain::block::{Block, BlockHeaderHash};
+use zebra_chain::types::BlockHeight;
+
+struct ChainVerifier<BV, S> {
+    /// The underlying `BlockVerifier`, possibly wrapped in other services.
+    block_verifier: BV,
+
+    /// The underlying `CheckpointVerifier`, wrapped in a buffer, so we can
+    /// clone and share it with futures.
+    checkpoint_verifier: Buffer<CheckpointVerifier, Arc<Block>>,
```

Comment:
I can't type erase the checkpoint verifier, because I need to access its maximum height:
https://github.com/ZcashFoundation/zebra/pull/691/files#diff-9ce8cedb3963941e6cffe1a307d9dafcR165-R167

I could add a trait `MaxHeight` for the `max_height` function, and constrain the generic service to implement that trait.

I would also need to generically impl `MaxHeight` for any service `S` that wraps a service `SMH: MaxHeight`. Otherwise, we couldn't actually wrap these services.

I'm not sure exactly how to do all these things. Do you think they're worth it? If you do, I'll open another ticket for that refactor.

I also think @hdevalence wanted to remove the generic `init` functions for internal services in `zebra-consensus`. If we just expose `ChainVerifier` publicly, then we could remove `init` for `BlockVerifier`, and most of the other services.

Reply:
My impression was that in cases where we need access to internals of a service @hdevalence prefers that we add a new `Request`/`Response` pair rather than adding fn's but if that's not the case I have absolutely no objections.

Reply:
I don't have a strong opinion here, I'm happy to add a Request enum with MaxHeight and Verify variants.

MaxHeight is constant, should we call it once and store it in the ChainVerifier?
It seems a bit wasteful to call it for every block.

Reply:
@yaahc the `Request`/`Response` pair is because a `Service` is supposed to be a function, not an object (i.e., there's only call, not other methods). But here I don't think that we need to do that, @teor2345's suggestion to get the max height once and save it sounds good.